components/brave_wallet/browser/eth_allowance_manager.cc

```diff
@@ -0,0 +1,409 @@
+    : task_id(taskid), account_address_(account_address), chain_id_(chain_id) {}
+EthAllowanceManager::EthAllowanceTask::~EthAllowanceTask() = default;
+void EthAllowanceManager::EthAllowanceTask::SetResults(
+    const uint256_t& latest_block_number,
+    std::vector<mojom::AllowanceInfoPtr> alwns) {
+  latest_block_number_ = latest_block_number;
+  allowances_ = std::move(alwns);
+  MarkComplete();
+}
+void EthAllowanceManager::EthAllowanceTask::MarkComplete() {
+  is_completed_ = true;
+}
+
+EthAllowanceManager::EthAllowanceManager(JsonRpcService* json_rpc_service,
+                                         KeyringService* keyring_service,
+                                         PrefService* prefs)
+    : json_rpc_service_(json_rpc_service),
+      keyring_service_(keyring_service),
+      prefs_(prefs) {}
+
+EthAllowanceManager::~EthAllowanceManager() = default;
+
+// Start of the allowance discovering operation.
+void EthAllowanceManager::DiscoverEthAllowancesOnAllSupportedChains(
+    ResultCallback callback) {
+  if (!discover_eth_allowance_callbacks_.empty()) {
+    discover_eth_allowance_callbacks_.push_back(std::move(callback));
+    return;
+  }
+  discover_eth_allowance_callbacks_.push_back(std::move(callback));
+
+  const auto keyring_info =
+      keyring_service_->GetKeyringInfoSync(mojom::kDefaultKeyringId);
+  std::vector<std::string> account_addresses;
+  for (const auto& account_info : keyring_info->account_infos) {
+    account_addresses.push_back(account_info->address);
+  }
+
+  if (account_addresses.empty()) {
+    OnDiscoverEthAllowancesCompleted({});
+    return;
+  }
+
+  const auto token_list_map =
+      BlockchainRegistry::GetInstance()->GetEthTokenListMap(
+          GetChainIdsForAlowanceDiscovering());
+
+  if (!allowance_discovery_tasks_.empty()) {
+    allowance_discovery_tasks_.clear();
+    get_block_tasks_ = 0;
+  }
+  for (const auto& [chain_id, token_list] : token_list_map) {
+    base::Value::List contract_addresses;
+    for (const auto& token : token_list) {
+      contract_addresses.Append(token->contract_address);
+    }
+
+    if (contract_addresses.empty()) {
+      continue;
+    }
+    auto internal_callback = base::BindOnce(
+        &EthAllowanceManager::OnGetCurrentBlock, weak_ptr_factory_.GetWeakPtr(),
+        chain_id, std::move(contract_addresses), account_addresses);
+    get_block_tasks_++;
+    json_rpc_service_->GetBlockNumber(chain_id, std::move(internal_callback));
+  }
+}
+
+void EthAllowanceManager::OnGetCurrentBlock(
+    const std::string& chain_id,
+    base::Value::List contract_addresses,
+    const std::vector<std::string>& account_addresses,
+    uint256_t block_num,
+    mojom::ProviderError error,
+    const std::string& error_message) {
+  const std::string latest_block_num_hex(error == mojom::ProviderError::kSuccess
+                                             ? Uint256ValueToHex(block_num)
+                                             : kLatestBlock);
```

Comment:
imho in case of any error we should just abort whole discovery

Reply:
added correct handling, i.e. aborting discovery in that case